<?xml version="1.0"?>
<launch>
  <!-- utm arguments -->
  <arg name="utm_zone" default="52" />
  <arg name="hemisphere" default="North" />
    <!-- k-city-->
  <arg name="gps_offset_x" default="302533.174487" />
  <arg name="gps_offset_y" default="4124215.34631" />
  <arg name="gps_offset_z" default="6.874900" />

  <!-- gps & imu arguments -->
  <arg name="imu_topic" default="vectornav/IMU" />
  <arg name="window_size" default="2" />
  
  <!-- lidar filtering arguments -->
  <arg name="laser_topic" default="/velodyne_points" />
  <arg name="scan_line" default="32" />
  <arg name="max_distance" default="150.0" />
  <arg name="min_distance" default="10.0" />
  <arg name="vertical_resolution" default="0.33" />
  <arg name="filtering_leaf_size" default="1.5" />
    <!-- <arg name="pcd_map_name" default="dlo_map.pcd" /> -->
  <arg name="robot_x_min" default="-2.0" />
  <arg name="robot_x_max" default="2.0" />
  <arg name="robot_y_min" default="-0.1" />
  <arg name="robot_y_max" default="0.1" />
  <arg name="robot_z_min" default="-1.0" />
  <arg name="robot_z_max" default="1.0" />
  <!-- ndt arguments -->
  <!-- <arg name="init_x" default="0.0" />
  <arg name="init_y" default="0.0" />
  <arg name="init_z" default="0.0" /> -->
  <arg name="init_x" default="-78.956013" />
  <arg name="init_y" default="-514.205941" />
  <arg name="init_z" default="0.0" />
  <arg name="init_rotation" default="0.96" />
  <arg name="map_resolution" default="1.5" />
  <arg name="map_path" default="$(find lidar_localizer)/map/" />
  <arg name="map_name" default="leaf_0.2.pcd" />
    <!-- submap_select 1 is k-nearest search, 0 is radius search-->
  <arg name="submap_select" default="1" />
  <arg name="kdtree_search_radius" default="200.0" />
  <arg name="ndt_near_points" default="400000" />
  <arg name="ndt_max_iteration" default="1000" />
  <arg name="ndt_max_threads" default="12" />
    <!-- transform matrix info if needed-->
  <arg name="rotation_theta" default="0.0" />
  <!-- <arg name="translation_x" default="-78.956013" />
  <arg name="translation_y" default="-514.205941" />
  <arg name="translation_z" default="6.874900" /> -->
  <arg name="translation_x" default="0.0" />
  <arg name="translation_y" default="0.0" />
  <arg name="translation_z" default="-6.874900" />


  <!-- rviz -->
  <arg name="rviz" default="true" />
  
  <!-- if using bagfile, set use_sim_time true-->
  <param name="/use_sim_time" value="true" />
  
  <!-- utm parameters-->
  <param name="utm_zone" type="int" value="$(arg utm_zone)" />
  <param name="hemisphere" type="string" value="$(arg hemisphere)" />
  <param name="gps_offset_x" type="double" value="$(arg gps_offset_x)" />
  <param name="gps_offset_y" type="double" value="$(arg gps_offset_y)" />
  <param name="gps_offset_z" type="double" value="$(arg gps_offset_z)" />
  
  <!-- gps & imu parameters-->
  <param name="imu_topic" type="string" value="$(arg imu_topic)" />
  <param name="window_size" type="int" value="$(arg window_size)" />

  <!-- lidar filtering parameters -->
  <param name="laser_topic" type="string" value="$(arg laser_topic)" />
  <param name="scan_line" type="int" value="$(arg scan_line)" />
  <param name="max_distance" type="double" value="$(arg max_distance)" />
  <param name="min_distance" type="double" value="$(arg min_distance)" />
  <param name="vertical_angle" type="double" value="$(arg vertical_resolution)" />
  <param name="filtering_leaf_size" type="double" value="$(arg filtering_leaf_size)" />
  <param name="robot_x_min" type="double" value="$(arg robot_x_min)" />
  <param name="robot_x_max" type="double" value="$(arg robot_x_max)" />
  <param name="robot_y_min" type="double" value="$(arg robot_y_min)" />
  <param name="robot_y_max" type="double" value="$(arg robot_y_max)" />
  <param name="robot_z_min" type="double" value="$(arg robot_z_min)" />
  <param name="robot_z_max" type="double" value="$(arg robot_z_max)" />

  <!-- ndt parameters -->
  <param name="init_x" type="double" value="$(arg init_x)" />
  <param name="init_y" type="double" value="$(arg init_y)" />
  <param name="init_z" type="double" value="$(arg init_z)" />
  <param name="init_rotation" type="double" value="$(arg init_rotation)" />
  <param name="map_resolution" type="double" value="$(arg map_resolution)" />
  <param name="map_path" type="string" value="$(arg map_path)" />
  <param name="map_name" type="string" value="$(arg map_name)" />
  <param name="submap_select" type="bool" value="$(arg submap_select)" />
  <param name="kdtree_search_radius" type="double" value="$(arg kdtree_search_radius)" />
  <param name="ndt_near_points" type="int" value="$(arg ndt_near_points)" />
  <param name="ndt_max_iteration" type="int" value="$(arg ndt_max_iteration)" />
  <param name="ndt_max_threads" type="int" value="$(arg ndt_max_threads)" />
  <param name="rotation_theta" type="double" value="$(arg rotation_theta)" />
  <param name="translation_x" type="double" value="$(arg translation_x)" />
  <param name="translation_y" type="double" value="$(arg translation_y)" />
  <param name="translation_z" type="double" value="$(arg translation_z)" />
  
  <!-- node -->
  <node pkg="lidar_localizer" type="gps_to_utm_node" name="gps_to_utm_node" output="screen">
  </node>
  <node pkg="lidar_localizer" type="lidar_processing_node" name="lidar_processing_node" output="screen">
  </node>
  <node pkg="lidar_localizer" type="ndt_matching_node" name="ndt_matching_node" output="screen">
  </node>
  <node pkg="lidar_localizer" type="gps_imu_node" name="gps_imu_node" output="log">
  </node>

  <node pkg="map_server" name="map_server" type="map_server" args="$(find lidar_localizer)/map/global_map.yaml"/>
  <node pkg="tf" type="static_transform_publisher" name="map2velodyne" args="0 0 0 0 0 0 /base_link /velodyne 10" />
  <node pkg="tf" type="static_transform_publisher" name="map2imu" args="0 0 0 0 0 0 /base_link /vectornav 10" />

  <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="trajectory_server_loam" ns="odom" >
    <param name="/target_frame_name" value="map" />
    <param name="/source_frame_name" value="base_link" />
    <param name="/trajectory_update_rate" value="10.0" />
    <param name="/trajectory_publish_rate" value="10.0" />
  </node>

  <group if="$(arg rviz)">
    <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find lidar_localizer)/rviz/ndt.rviz" />
  </group>

</launch>
